feat(dashboards): Add Edit Access Selector to Dashbaords

harshithadurai opened the pull request:
#79826 

- Add Edit Access Selector Button to the UI, and hook up the dashboard detail endpoints to update Edit permissions.

<img width="1280" alt="Screenshot 2024-10-28 at 10 54 39 AM" src="https://github.com/user-attachments/assets/eff9249d-4fad-4356-b009-45c0b7bd1ec1">

<img height="200" alt="Screenshot 2024-10-28 at 10 55 31 AM" src="https://github.com/user-attachments/assets/7bfe68c4-eb8d-4fe3-afcc-74f92b76ff2e">

<img height="200" alt="Screenshot 2024-10-28 at 10 55 44 AM" src="https://github.com/user-attachments/assets/f11c7d00-33f4-4b84-bbfd-8b17b1204418">

<br>
<br>

- Disables the ‘Edit Dashboard’ and ‘Add Widget’ buttons when the user does not have perms to Edit the Dashboard.

<img width="1252" alt="Screenshot 2024-10-28 at 10 59 30 AM" src="https://github.com/user-attachments/assets/75d16c5a-9ab9-4f0f-a5ce-059168de60a6">

Commit:
edit access selector component

harshithadurai committed:
Merge branch 'master' into harshi/feat/dashboard-creator-only-edit-permissions-frontend

Commit:
pass down dashboardperms

harshithadurai committed:
Merge branch 'master' into harshi/feat/dashboard-creator-only-edit-permissions-frontend

Commit:
api request, component changes

harshithadurai committed:
Merge branch 'master' into harshi/feat/dashboard-creator-only-edit-permissions-frontend

Commit:
test fix

Commit:
grey out 'add widget' and 'edit dashboard' buttons

harshithadurai committed:
Merge branch 'master' into harshi/feat/dashboard-creator-only-edit-permissions-frontend

Commit:
details.spec test

harshithadurai commented on static/app/views/dashboards/editAccessSelector.tsx:
```diff
@@ -0,0 +1,196 @@
+ * Dropdown multiselect button to enable selective Dashboard editing access to
+ * specific users and teams
+ */
+function EditAccessSelector({dashboard, onChangeEditAccess}: EditAccessSelectorProps) {
```
Here, `dashboard` isn't a state variable. This means that when the perms are updated, it doesn't instantly update the overall dashboard's state. However, I 'locally' manage the perms state within the component (line 35). 

I figured this would be okay since only the creator will be editing perms and it doesn't need to instantly reflect on the dashboard. Is this okay?

Commit:
tests

Commit:
add more tests

harshithadurai committed:
Merge branch 'master' into harshi/feat/dashboard-creator-only-edit-permissions-frontend

narsaynorath commented on static/app/views/dashboards/detail.tsx:
```diff
@@ -167,6 +170,23 @@ export function handleUpdateDashboardSplit({
+  currentUser: User,
+  organization: Organization
+): boolean {
+  if (organization.features.includes('dashboards-edit-access')) {
+    if (!dashboard.permissions) {
+      return true;
+    }
+    return dashboard.permissions.isCreatorOnlyEditable
+      ? dashboard.createdBy?.id === currentUser.id
+      : !dashboard.permissions.isCreatorOnlyEditable;
+  }
+  return true;
```
This is a nit pick so it's optional feedback, but nesting adds a bit of complexity so it's preferable to exit early. It also allows us to combine the conditions together here

```suggestion
  if (!organization.features.includes('dashboards-edit-access') || !dashboard.permissions) {
    return true;
  }
  return dashboard.permissions.isCreatorOnlyEditable
    ? dashboard.createdBy?.id === currentUser.id
    : !dashboard.permissions.isCreatorOnlyEditable;
```

narsaynorath commented on static/app/views/dashboards/detail.tsx:
```diff
@@ -594,6 +614,24 @@ class DashboardDetail extends Component<Props, State> {
+    updateDashboard(api, organization.slug, dashboardCopy).then(
+      (newDashboard: DashboardDetails) => {
+        addSuccessMessage(t('Dashboard Edit Access updated.'));
+        return newDashboard;
```
In other places of this file, e.g. `onCommit` we use `onDashboardUpdate` to set the dashboard in the parent, does this solve the issue of passing the state down properly? I don't think you'll need to update the state because there shouldn't be a "modified dashboard" when this accessible

narsaynorath commented on static/app/views/dashboards/editAccessSelector.tsx:
```diff
@@ -0,0 +1,198 @@
+              : dashboardCreator?.email || currentUser.email}
+          </StyledDisplayName>
+        }
+        displayEmail="Creator"
```
All of our strings need to be wrapped with `t()` or `tct()` depending on the use case, this is for marking text as translatable.

```suggestion
        displayEmail={t("Creator")}
```

narsaynorath commented on static/app/views/dashboards/editAccessSelector.tsx:
```diff
@@ -0,0 +1,198 @@
+        displayName={
+          <StyledDisplayName>
+            {dashboardCreator?.id === currentUser.id
+              ? `You (${currentUser.email})`
```
In this case we should probably use `tct`, it would be applied like so

```suggestion
              ? tct("You ([email])", {email: currentUser.email})
```

narsaynorath commented on static/app/views/dashboards/editAccessSelector.tsx:
```diff
@@ -0,0 +1,198 @@
+          <StyledDisplayName>
+            {dashboardCreator?.id === currentUser.id
+              ? `You (${currentUser.email})`
+              : dashboardCreator?.email || currentUser.email}
```
I'm not exactly sure that falling back to the `currentUser.email` makes sense here. If this is supposed to cover the create state (i.e. the scenario where the dashboard creator email is undefined), don't we want to display the current user email the same as `You (${currentUser.email})`?

narsaynorath commented on static/app/views/dashboards/editAccessSelector.tsx:
```diff
@@ -0,0 +1,198 @@
+  // Avatars/Badges in the Edit Selector Button
+  const triggerAvatars =
+    isEverythingSelected || !dashboardCreator ? (
+      <StyledBadge text={'All'} />
```
```suggestion
      <StyledBadge text={t('All')} />
```

narsaynorath commented on static/app/views/dashboards/editAccessSelector.tsx:
```diff
@@ -0,0 +1,198 @@
+      options: [
+        {
+          value: '_everyone',
+          label: 'Everyone',
```
```suggestion
          label: t('Everyone'),
```

narsaynorath commented on static/app/views/dashboards/editAccessSelector.tsx:
```diff
@@ -0,0 +1,198 @@
+  );
+
+  return dashboardCreator?.id !== currentUser.id ? (
+    <Tooltip title={'Only Dashboard Creator may change Edit Access'}>
```
```suggestion
    <Tooltip title={t('Only Dashboard Creator may change Edit Access')}>
```

narsaynorath commented on static/app/views/dashboards/editAccessSelector.tsx:
```diff
@@ -0,0 +1,198 @@
+        <React.Fragment key="edit-access-label">{t('Edit Access:')}</React.Fragment>,
+        <React.Fragment key="trigger-avatars">{triggerAvatars}</React.Fragment>,
+      ]}
+      searchPlaceholder="Search Teams"
```
```suggestion
      searchPlaceholder={t("Search Teams")}
```

narsaynorath commented on static/app/views/dashboards/editAccessSelector.tsx:
```diff
@@ -0,0 +1,198 @@
+      options={dropdownOptions}
+      value={selectedOptions}
+      triggerLabel={[
+        <React.Fragment key="edit-access-label">{t('Edit Access:')}</React.Fragment>,
```
Typically we import `Fragment` directly from `react` so we don't do `React.Fragment`. Also, `Fragment` is typically used to wrap multiple elements that belong together but don't exactly have a root, since they're just wrapping a single thing, you should be able to do `triggerLabel={[t('Edit Access:'), triggerAvatars]}`

narsaynorath commented on static/app/views/dashboards/editAccessSelector.tsx:
```diff
@@ -0,0 +1,198 @@
+
+  const [newDashboardPermissions, setNewDashboardPermissions] = useState<
+    DashboardPermissions | undefined
+  >(dashboard.permissions ? structuredClone(dashboard.permissions) : undefined);
```
Oh cool! I didn't know about `structuredClone`

narsaynorath reviewed (commented):
Some comments to start with! I noticed there were also tests that were commented out, but I think that's probably what we'll chat about later today

harshithadurai committed:
Update static/app/views/dashboards/detail.tsx

Co-authored-by: Nar Saynorath <nar.saynorath@sentry.io>

harshithadurai committed:
Update static/app/views/dashboards/editAccessSelector.tsx

change tooltip

Co-authored-by: Nar Saynorath <nar.saynorath@sentry.io>

harshithadurai committed:
Update static/app/views/dashboards/editAccessSelector.tsx

add 't' to string

Co-authored-by: Nar Saynorath <nar.saynorath@sentry.io>

getsantry[bot] committed:
:hammer_and_wrench: apply pre-commit fixes

harshithadurai committed:
Update static/app/views/dashboards/editAccessSelector.tsx

add 't' to string

Co-authored-by: Nar Saynorath <nar.saynorath@sentry.io>

harshithadurai committed:
Update static/app/views/dashboards/editAccessSelector.tsx

add 'tct' to email

Co-authored-by: Nar Saynorath <nar.saynorath@sentry.io>

harshithadurai committed:
Update static/app/views/dashboards/editAccessSelector.tsx

wrap string in t()

Co-authored-by: Nar Saynorath <nar.saynorath@sentry.io>

getsantry[bot] committed:
:hammer_and_wrench: apply pre-commit fixes

Commit:
un-nest if-condition

Commit:
update import

Commit:
fix "All" Badge

harshithadurai committed:
Merge branch 'master' into harshi/feat/dashboard-creator-only-edit-permissions-frontend

Commit:
update fragment use and add key in right place

Commit:
test

Commit:
add test

Commit:
update test

narsaynorath committed:
fix test

- initializes the page filter state so it doesn't block loading on
  individual page run
- sets the user fixture in the dashboard as created by

Commit:
fix tests

Commit:
test typo

Commit:
disable creator option in dropdown

narsaynorath commented on static/app/views/dashboards/editAccessSelector.tsx:
```diff
@@ -0,0 +1,190 @@
+  ];
+
+  // Handles state change when dropdown options are selected
+  const onSelectOptions = newSelectedOptions => {
+    const newSelectedValues = newSelectedOptions.map(
+      (option: {value: string}) => option.value
+    );
+    isEverythingSelected = false;
+    if (newSelectedValues.includes('_everyone')) {
+      isEverythingSelected = true;
+      setselectedOptions(['_everyone', '_creator', ...teamIds]);
+      if (newDashboardPermissions) {
+        newDashboardPermissions.isCreatorOnlyEditable = false;
+      }
+    } else if (!newSelectedValues.includes('_everyone')) {
+      setselectedOptions(['_creator']);
+      if (newDashboardPermissions === undefined) {
+        // When the dashboard does not have a permissions model associated
+        setNewDashboardPermissions({isCreatorOnlyEditable: true});
+      } else {
+        newDashboardPermissions.isCreatorOnlyEditable = true;
+      }
+    }
+  };
```
Do we need to store the newDashboardPermissions every time a change happens? Can we just store the selected options and then when we trigger `onClose` we can create the new dashboard permissions for the request adhoc?

That way we can get rid of a state variable for `newDashboardPermissions`

harshithadurai replied:
Removed the state hook for `newDashboardPermissions`, and added [this function](https://github.com/getsentry/sentry/blob/48a638d9ced60c8a02e4d09a6fcb62292e8e94b3/static/app/views/dashboards/editAccessSelector.tsx#L126)

narsaynorath reviewed (commented):
Okay! Not too many comments left, but I wanted to see if we could simplify the `onChange` logic. It's looking a bit large for what it needs to do and I think we only need `newDashboardPermissions` when we close the compact select.

Commit:
update the ewya newperms are set, fix tests

narsaynorath reviewed (commented):
Just some last suggestions :pray:

Commit:
handle default state perms

Commit:
update creator email when dashboard is being created

Commit:
remove iseverythingselected

Commit:
remove teams from dropdown

harshithadurai commented on static/app/views/dashboards/editAccessSelector.tsx:
```diff
@@ -0,0 +1,203 @@
+  );
+
+  let isEverythingSelected =
+    selectedOptions.length === ['_everyone', '_creator', ...teamIds].length;
```
Makes sense. I removed the `isEverythingSelected` variable entirely, since it was essentially just being used to display the 'All' Badge in the dropdown button:
```
  // Avatars/Badges in the Edit Selector Button
  const triggerAvatars =
    selectedOptions.includes('_everyone') || !dashboardCreator ? (
      <StyledBadge key="_all" text={'All'} />
    ) : (
      <StyledAvatarList key="avatar-list" users={[dashboardCreator]} avatarSize={25} />
    );
```
(from [this file](https://github.com/getsentry/sentry/blob/c3444c2e0f3b2e469d00809e0f64b00f57e746e6/static/app/views/dashboards/editAccessSelector.tsx#L73))

narsaynorath reviewed (approved):
🎉  Awesome!
